File: packages/ruby/lib/identifold/storage.rb
module Identifold
  module Storage
    ReferenceReservation = Data.define(:machine_id, :namespace, :reference)
    ReferenceMapping = Data.define(:machine_id, :namespace)
    SequenceAllocationRequest = Data.define(
      :machine_id,
      :namespace,
      :reference_prefix,
      :scope,
      :width
    )
    Adapter = Data.define(:reference_store, :reference_lookup, :sequence_allocator)
  end
end

File: packages/go/storage/storage.go
package storage

import "context"

type ReferenceReservation struct {
	MachineID string
	Namespace string
	Reference string
}

type ReferenceMapping struct {
	MachineID string
	Namespace string
}

type SequenceAllocationRequest struct {
	MachineID       string
	Namespace       string
	ReferencePrefix string
	Scope           *string
	Width           uint8
}

type Adapter interface {
	Reserve(context.Context, ReferenceReservation) (bool, error)
	Resolve(context.Context, string, string) (*ReferenceMapping, error)
	Allocate(context.Context, SequenceAllocationRequest) (uint64, error)
}
